pkg/ccl/backupccl/restore_online.go

```diff
@@ -171,11 +172,40 @@ func sendAddRemoteSSTWorker(
 		}
 
 		for entry := range restoreSpanEntriesCh {
+			// If we're restoring a tenant, we need to move the end key of any spans
+			// that ended at the exclusive end of the tenant span, i.e. the start key
+			// of the next tenant ID, into the prefix of this tenant span instead so
+			// that that end key can be rewritten to the restoring tenant ID's prefix.
+			// We do this be replacing any key equal to TenantSpan.EndKey (tID+1) with
+			// tenantSpan.Key followed by keyMax, since keyMax sorts able all table
+			// keys, including within the subspace of a tenant span.
+			if entry.ElidedPrefix == execinfrapb.ElidePrefix_Tenant {
+				_, id, err := keys.DecodeTenantPrefix(entry.Span.Key)
+				if err != nil {
+					return err
+				}
+				if tSpan := keys.MakeTenantSpan(id); entry.Span.EndKey.Equal(tSpan.EndKey) {
+					entry.Span.EndKey = append(tSpan.Key, keys.MaxKey...)
+				}
+			}
 			firstSplitDone := false
 			if err := assertCommonPrefix(entry.Span, entry.ElidedPrefix); err != nil {
 				return err
 			}
 			for _, file := range entry.Files {
+				if entry.ElidedPrefix == execinfrapb.ElidePrefix_Tenant {
+					_, id, err := keys.DecodeTenantPrefix(file.BackupFileEntrySpan.Key)
+					if err != nil {
+						return err
+					}
+					if tSpan := keys.MakeTenantSpan(id); file.BackupFileEntrySpan.EndKey.Equal(tSpan.EndKey) {
+						file.BackupFileEntrySpan.EndKey = append(tSpan.Key, keys.MaxKey...)
+					}
+				}
+				if err := assertCommonPrefix(file.BackupFileEntrySpan, entry.ElidedPrefix); err != nil {
```

Comment:
nit: we could remove the `assertCommonPrefix()` call in `rewriteSpan()`, as we now check the file and the span entry before intersection.